core/application/impl/app_config_impl.cpp

```diff
@@ -0,0 +1,332 @@
+    load_str(val, "rpc_http_host", rpc_http_host_);
+    load_u16(val, "rpc_http_port", rpc_http_port_);
+    load_str(val, "rpc_ws_host", rpc_ws_host_);
+    load_u16(val, "rpc_ws_port", rpc_ws_port_);
+  }
+
+  void AppConfigurationImpl::parse_additional_segment(rapidjson::Value &val) {
+    load_bool(val, "single_finalizing_node", is_only_finalizing_);
+  }
+
+  void AppConfigurationImpl::validate_config(
+      AppConfiguration::LoadScheme scheme) {
+    if (genesis_path_.empty()) {
+      logger_->error("Node configuration must contain 'genesis' option.");
+      exit(EXIT_FAILURE);
+    }
+
+    if (leveldb_path_.empty()) {
+      logger_->error("Node configuration must contain 'leveldb_path' option.");
+      exit(EXIT_FAILURE);
+    }
+
+    const auto need_keystore =
+        (AppConfiguration::LoadScheme::kBlockProducing == scheme)
+        || (AppConfiguration::LoadScheme::kValidating == scheme);
+
+    if (!(!need_keystore != !keystore_path_.empty())) {
+      logger_->error("Node configuration must contain 'keystore_path' option.");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  void AppConfigurationImpl::read_config_from_file(
+      const std::string &filepath) {
+    assert(!filepath.empty());
+
+    /// TODO: make handler calls via lambda-calls, remove member-function ptrs
+    /// (iceseer)
+    RegisteredHandlers handlers[] = {
+        {"general", &AppConfigurationImpl::parse_general_segment},
+        {"blockchain", &AppConfigurationImpl::parse_blockchain_segment},
+        {"storage", &AppConfigurationImpl::parse_storage_segment},
+        {"authority", &AppConfigurationImpl::parse_authority_segment},
+        {"network", &AppConfigurationImpl::parse_network_segment},
+        {"additional", &AppConfigurationImpl::parse_additional_segment},
+    };
+
+    auto file = open_file(filepath);
+    if (!file) {
+      logger_->error("Configuration file path is invalid: {}", filepath);
+      exit(EXIT_FAILURE);
+    }
+
+    using FileReadStream = rapidjson::FileReadStream;
+    using Document = rapidjson::Document;
+
+    char buffer_size[1024];
+    FileReadStream input_stream(file.get(), buffer_size, sizeof(buffer_size));
+
+    Document document;
+    document.ParseStream(input_stream);
+
+    for (auto &handler : handlers) {
+      auto it = document.FindMember(handler.segment_name);
+      if (document.MemberEnd() != it) {
+        (this->*handler.handler)(it->value);
+      }
+    }
+  }
+
+  boost::asio::ip::tcp::endpoint AppConfigurationImpl::get_endpoint_from(
+      std::string const &host, uint16_t port) {
+    boost::asio::ip::tcp::endpoint endpoint;
+    boost::system::error_code err;
+
+    endpoint.address(boost::asio::ip::address::from_string(host, err));
+    if (err.failed()) {
+      logger_->error("RPC address '{}' is invalid", host);
+      exit(EXIT_FAILURE);
+    }
+
+    if (port <= 0 || port >= 65535) {
+      logger_->error("RPC port '{}' is wrong", port);
+      exit(EXIT_FAILURE);
+    }
```

Comment:
This check may be late. =)

Reply:
Hmmm! Yep, it was the plan - to make validation in a single place after all config data manipulations! Thanks:)

Reply:
This check seems to be redundant. Argument port is ```uint16_t```...it will never fails this check.

Reply:
Yep. Except zero-port and usualy-reserved port 65535. =) Better to check it on another level (e.g., at parsing).

Reply:
I think 65535 is a dynamic port and it has no IANA-registered appointment.